### PR TITLE
Viite 746 roadaddress change import to road address

### PR DESCRIPTION
**Pull request by hestrompa**

### CODE REVIEW

File: digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala [hunk 1]
```diff
@@ -2157,4 +2164,205 @@ class ProjectServiceSpec  extends FunSuite with Matchers with BeforeAndAfter {
...
+        None, Some("bum"), 0L, 413424L, 0.0, 24.539, AgainstDigitizing, (None, None), false, Seq(Point(1024.0,2048.0), Point(1048.0, 2089.0)),
+        1L, LinkStatus.New, RoadType.PublicRoad, LinkGeomSource.SuravageLinkInterface, 24.539, 123L, 8L, Some(77447L)),
+      ProjectLink(1001L, 5L, 8L, Combined, Continuous, 982L, 1007L, Some(DateTime.now()),
+        projectDate, Some("bum"), 0L, 77447L, 0.0, 24.539, AgainstDigitizing, (None, None), false, Seq(Point(1024.0,2048.0), Point(1048.0, 2089.0)),
+        1L, LinkStatus.Terminated, RoadType.PublicRoad, LinkGeomSource.SuravageLinkInterface, 24.539, 123L, 8L, Some(413424L)))
+    val project = RoadAddressProject(1L, ProjectState.Sent2TR, "proj", "bum", DateTime.now(), "bum", projectDate.get,
+      DateTime.now(), "", Seq(), None, Some(8L))
+    val resultRA = projectService.createSplitRoadAddress(roadAddress, splitProjectlinks, project)
+    resultRA should have size (4)
+//    resultRA.foreach(ra =>
```
Comment (pedrosag) [inline digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala]: Verify if this code is needed please.

File: digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala [hunk 2]
```diff
@@ -2157,4 +2164,205 @@ class ProjectServiceSpec  extends FunSuite with Matchers with BeforeAndAfter {
...
+      ProjectLink(1000L, 5L, 8L, Combined, Continuous, 1112L, 1137L, projectDate,
+        None, Some("bum"), 0L, 413424L, 31.837, 56.376, TowardsDigitizing, (None, None), false, Seq(Point(1024.0,2048.0), Point(1048.0, 2089.0)),
+        1L, LinkStatus.New, RoadType.PublicRoad, LinkGeomSource.SuravageLinkInterface, 24.539, 123L, 8L, Some(77447L)),
+      ProjectLink(1001L, 5L, 8L, Combined, Continuous, 982L, 1007L, Some(DateTime.now()),
+        projectDate, Some("bum"), 0L, 77447L, 31.837, 56.376, TowardsDigitizing, (None, None), false, Seq(Point(1024.0,2048.0), Point(1048.0, 2089.0)),
+        1L, LinkStatus.Terminated, RoadType.PublicRoad, LinkGeomSource.NormalLinkInterface, 24.539, 123L, 8L, Some(413424L)))
+    val project = RoadAddressProject(1L, ProjectState.Sent2TR, "proj", "bum", DateTime.now(), "bum", projectDate.get,
+      DateTime.now(), "", Seq(), None, Some(8L))
+    val resultRA = projectService.createSplitRoadAddress(roadAddress, splitProjectlinks, project)
+    //    resultRA should have size (4)
```
Comment (pedrosag) [inline digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala]: Verify if this code is needed please.

File: digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala [hunk 3]
```diff
@@ -2157,4 +2164,205 @@ class ProjectServiceSpec  extends FunSuite with Matchers with BeforeAndAfter {
...
+      None, Some("bum"), 0L, 77452L, 0.0, 20.947, TowardsDigitizing, (None, None), false, Seq(Point(1021.05, 2048.0), Point(1024.0, 2048.0)),
+      1L, LinkStatus.New, RoadType.PublicRoad, LinkGeomSource.NormalLinkInterface, 20.947, 125L, 8L, None)
+    val vvhLinks = Seq(
+      toVVHRoadlink(roadAddressSplit), toVVHRoadlink(roadAddressTerm), toVVHRoadlink(roadAddressTrsnf), toVVHRoadlink(projectLinkNew),
+      toVVHRoadlinkS(splitProjectlinks.filter(_.linkId == 413424L)), toVVHRoadlinkS(splitProjectlinks.filter(_.linkId == 77447L))
+    )
+    val createdRA = projectService.convertToRoadAddress(splitProjectlinks, Seq(projectLinkTerm, projectLinkTrsnf), Seq(projectLinkNew),
+      vvhLinks.map(l => l.linkId -> l).toMap,
+      Seq(roadAddressSplit, roadAddressTerm, roadAddressTrsnf).map(ra => ra.id -> ra).toMap, project)
+//    createdRA.sortBy(_.endDate.map(_.getMillis).getOrElse(Long.MaxValue)).foreach(ra => {
```
Comment (pedrosag) [inline digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala]: Verify if this code is needed please.

File: digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala [hunk 4]
```diff
@@ -2157,4 +2164,205 @@ class ProjectServiceSpec  extends FunSuite with Matchers with BeforeAndAfter {
       updatedLinks.maxBy(_.endAddrMValue).id should be (updatedLinks.find(_.discontinuity.value == 2).get.id)
     }
   }
+
+  test("Calculate split road addresses with Unchanged + Term/New") {
+    val projectDate = Some(DateTime.parse("2021-01-01"))
+    val roadAddress = RoadAddress(123L, 5L, 8L, RoadType.PublicRoad, Combined, Continuous, 1080L, 1137L,
```
Comment (pedrosag) [inline digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectServiceSpec.scala]: Please be careful when using fixed numbers for ID's, try to use the best sequence for them:
VIITE_GENERAL_SEQ;
VIITE_PROJECT_SEQ;
LRM_POSITION_PRIMARY_KEY_SEQ;
PRIMARY_KEY_SEQ;

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala [hunk 5]
```diff
@@ -1120,105 +1121,150 @@ class ProjectService(roadAddressService: RoadAddressService, roadLinkService: Ro
     }
   }
 
+  def createSplitRoadAddress(roadAddress: RoadAddress, split: Seq[ProjectLink], project: RoadAddressProject): Seq[RoadAddress] = {
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: This method definitely needs unit tests.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala [hunk 6]
```diff
@@ -1120,105 +1121,150 @@ class ProjectService(roadAddressService: RoadAddressService, roadLinkService: Ro
...
+        case New =>
+          Seq(RoadAddress(NewRoadAddress, pl.roadNumber, pl.roadPartNumber, pl.roadType, pl.track,
+            pl.discontinuity, pl.startAddrMValue, pl.endAddrMValue, Some(project.startDate), None, Some(project.createdBy),
+            0L, pl.linkId, pl.startMValue, pl.endMValue, pl.sideCode, VVHClient.createVVHTimeStamp(), pl.calibrationPoints,
+            floating = false, pl.geometry, pl.linkGeomSource, pl.ely, terminated = false))
+        case Transfer =>
+          val (startAddr, endAddr, startM, endM) = transferValues(split.find(_.status == Terminated))
+          Seq(
+            // Transferred part, original values
+            roadAddress.copy(id = NewRoadAddress, endDate = Some(project.startDate),
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: Doesn't look right to me: picks terminated values but does not set termination; doesn't prevent terminated case.

Comment (hestrompa) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: Am I missing something? The terminated case is here:
case Terminated =>
          Seq(roadAddress.copy(id = NewRoadAddress, startAddrMValue = pl.startAddrMValue,
            endAddrMValue = pl.endAddrMValue, endDate = Some(project.startDate), startMValue = pl.startMValue, endMValue = pl.endMValue,
            linkId = pl.linkId, terminated = true, modifiedBy = Some(project.createdBy)))

This contains the terminated values and set termination = true.

Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: I'll do a unit test and I'll try to check if it does handle all cases properly.

Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: I read the code wrong, transferValues picks the non-terminated part for terminated argument.

**Pull request merged**